### PR TITLE
Use !! in AddParameterCheck refactoring

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #36024 

@CyrusNajmabadi for review

### CODE REVIEW

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:97]: GetRequiredSyntaxTree (and move null check)

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:104]: We need to track this somehow. I think Julien has a apattern on how to reference a lang version without using 'preview'

Comment (RikkiGibson) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:104]: Sure. I was going to create a tracking issue to revisit usages of LangVersion.Preview before release. But maybe @jcouv's method would be better.

Comment (RikkiGibson) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:104]: I see that #54359 added `LanguageVersionFacts.CSharpNext`. But it seems it's not accessible from `Features`.

Comment (jcouv) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:100]: You can add something similar to the IDE layer. If so, please add a comment next to CSharpNext so that we remember to fix both when adding a new language version.

File: src/Workspaces/CSharpTest/Formatting/FormattingTests.cs
Comment (jcouv) [inline src/Workspaces/CSharpTest/Formatting/FormattingTests.cs:5780]: Don't forget to also test/handle the syntax normalize (SyntaxNormalizerTests.cs) which generally needs similar fixing.

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:42]: Please add a test where we're starting with `string s!!` (fix should not be offered)

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (jcouv) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:100]: Should we also check whether the parameter already has `!!`?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/TokenBasedFormattingRule.cs
Comment (jcouv) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/TokenBasedFormattingRule.cs:486]: Consider also adding a test to `SyntacticClassifierTests.cs ` and `SemanticQuickInfoSourceTests.cs`. This way, the PR covers all IDE basics (formatting, classifier, quick info, since completion is N/A).

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (Youssef1313) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:106]: The node was already calculated in:

https://github.com/dotnet/roslyn/blob/d959aaf644ccc7f23f011cc4e8cadcf5b50133c9/src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs#L72

Would it be easy to pass it around instead of re-calculating?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/TokenBasedFormattingRule.cs
Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/TokenBasedFormattingRule.cs:482]:
```suggestion
            if (currentToken.IsKind(SyntaxKind.ExclamationExclamationToken) &&
```

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (Youssef1313) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:100]: > Should we also check whether the parameter already has `!!`?

I agree with that. This should be added, probably in `ParameterValidForNullCheck` around
https://github.com/dotnet/roslyn/blob/0c54df55b42fa98c81d2b88d9be2f5012ca28721/src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs#L266

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:106]: yes.  i'd prefer this as well, so all features have a consistent view of what syntax they're operating on (may be relevant in error cases).

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs:333]: Nit.  Consider:

```suggestion
            // First see if we can adopt the '!!' parameter null checking syntax.
            // Then see if we can convert a statement of the form "this.s = s" into "this.s = s ?? throw ...".
            var documentOpt =
                await TryAddNullCheckToParameterDeclarationAsync(document, parameter, cancellationToken).ConfigureAwait(false) ??
                await TryAddNullCheckToAssignmentAsync(document, parameter, blockStatementOpt, cancellationToken).ConfigureAwait(false);
```

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs:333]: (up to you on this though).

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:311]: For the cases where the test is the same other than changing language version, consider using a xUnit theory instead of a fact to avoid all the extra duplication.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (jasonmalinowski) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:43]: Remove the "Opt" suffix now that it's annotated?

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs
Comment (jasonmalinowski) [inline src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs:326]: Remove the Opt suffix now?

**Review by jasonmalinowski — APPROVED** [review]
Some comments from me, but wait until @CyrusNajmabadi signs off.

File: src/Workspaces/CSharpTest/Formatting/FormattingTests.cs
Comment (RikkiGibson) [reply src/Workspaces/CSharpTest/Formatting/FormattingTests.cs:5780]: I didn't see a similar test in SyntaxNormalizerTests for the `!` operator. Didn't we have to do something similar with that for formatting?

**Review by jcouv — APPROVED** [review]
Compiler-side (comments-only change) LGTM Thanks (iteration 4)

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (RikkiGibson) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs]: Done.

File: src/Workspaces/CSharpTest/Formatting/FormattingTests.cs
Comment (RikkiGibson) [reply src/Workspaces/CSharpTest/Formatting/FormattingTests.cs:5780]: It wasn't clear to me what needs to be done here.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (RikkiGibson) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs]: Done.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/TokenBasedFormattingRule.cs
Comment (RikkiGibson) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/TokenBasedFormattingRule.cs]: Done.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (RikkiGibson) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs]: Done.

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs
Comment (RikkiGibson) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: Done.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (Youssef1313) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:97]: `parameter` seems to be no longer used?

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs
Comment (RikkiGibson) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: It would read more nicely this way, but I prefer the current way so that it's easy to set breakpoints on each call.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (RikkiGibson) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs]: Good catch, thanks.

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:73]: i don't think you need this line, right?

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:98]: i don't think this is needed.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:107]: so, in general, we would far prefer the CT just be passed along and used here.  that way we have no dependencies/assumptions we have to fulfill.

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs:181]: this worries me.  were we blindly casting before?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs
Comment (RikkiGibson) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs:181]: It appears `parameterNodes` contains a `SeparatedSyntaxList<ParameterNode>` which was converted to an `IReadOnlyList<SyntaxNode>` within [`SyntaxGenerator.GetParameters`](https://github.com/dotnet/roslyn/blob/6aa373259bd1f19096ccf38cc8028293831a64de/src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs#L2001).

I would be open to any refactoring/additional helpers which would make this code safer/more maintainable.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (RikkiGibson) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs]: Fixed.

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs
Comment (RikkiGibson) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: I changed to simply cast the list of parameter nodes which I think will make it more clear what exactly our assumption is.

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (RikkiGibson) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: Fixed

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (RikkiGibson) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: Fixed

**Comment by RikkiGibson** [comment]
Could I please get a second review on the relatively small (I think only comments) changes on the @dotnet/roslyn-compiler side?

**Review by cston — APPROVED** [review]
Compiler changes LGTM.

File: src/Compilers/Core/Portable/Symbols/IParameterSymbol.cs
Comment (Youssef1313) [inline src/Compilers/Core/Portable/Symbols/IParameterSymbol.cs:96]: nit: I don't know whether this should be `<c>` or not. This will render differently on docs.microsoft.com.

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:36]: Why is this `Preview` and not the new `CSharpNext`? I understand both are equivalent, but `CSharpNext` should help updating this to CSharp11 when released.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs
Comment (jaredpar) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:104]: I was a bit surprised to see this. My expectation is that we would let the operation proceed without syntax checks and let later diagnostics catch that. Is there any other place where we have syntax based ops that depend on `/langVersion`?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs:104]: this is very normal on the IDE side.  remember that the user is just saying they want to add a "null check", not that they want to add `!!`.  The IDE normally takes a stance with these types of operations on choosing the most idiomatic form that is legal for the version of the language the user is on.

That way users on C#10 will get the best option they can use, wihle those on 11 will get `!!`.  

If users explicitly write out the C#11 version, then our 'upgrade project' analyzer kicks in and will help them upgrade if they want.

File: src/Compilers/Core/Portable/Symbols/IParameterSymbol.cs
Comment (RikkiGibson) [reply src/Compilers/Core/Portable/Symbols/IParameterSymbol.cs:96]: Perhaps it would make sense to have a tracking issue to verify the rendering/appearance of documentation comments, and perform adjustments like the one you've suggested.

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (RikkiGibson) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:36]: It's an oversight. Thanks for catching it.

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:282]:
```suggestion
        [CombinatorialData]
```

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:329]:
```suggestion
        [CombinatorialData]
```

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs
Comment (RikkiGibson) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:282]: This ends up giving us unexpected diagnostics about use of partial methods in C# 2.

**Pull request merged**
